trio/_dtls.py

```diff
@@ -1167,12 +1178,16 @@ def __del__(self):
                 f"unclosed DTLS endpoint {self!r}", ResourceWarning, source=self
             )
 
-    def close(self):
+    def close(self) -> None:
         """Close this socket, and all associated DTLS connections.
 
         This object can also be used as a context manager.
 
         """
+        # Do nothing if this object was never fully constructed
+        if self.socket is None:  # pragma: no cover
+            return
```

Comment:
... I don't think this case is possible? Well not unless someone's doing something _really_ funky like overriding our `__del__` to call `.close()`. Which is like, no, I don't think people will be doing that.

I think an `assert` is good enough, though it would be nice to have `socket` typed as non-optional. Not exactly sure how to do that though.

Reply:
Adding a comment so it can be addressed in an upcoming typing PR. I don't want to gradually add more typing to this PR, requiring more reviews.